chain/index/api.go

```diff
@@ -0,0 +1,345 @@
+		return nil, xerrors.Errorf("failed to get tipset at height %d: %w", epoch, err)
+	}
+
+	// we need to take a write lock here so that back-filling does not race with real time chain indexing
+	si.writerLk.Lock()
+	defer si.writerLk.Unlock()
+
+	var isIndexEmpty bool
+	if err := si.stmts.isIndexEmptyStmt.QueryRowContext(ctx).Scan(&isIndexEmpty); err != nil {
+		return nil, xerrors.Errorf("failed to check if index is empty: %w", err)
+	}
+
+	// Canonical chain has a null round at the epoch -> return if index is empty otherwise validate that index also
+	// has a null round at this epoch i.e. it does not have anything indexed at all for this epoch
+	if expectedTs.Height() != epoch {
+		if isIndexEmpty {
+			return &types.IndexValidation{
+				Height:      epoch,
+				IsNullRound: true,
+			}, nil
+		}
+		// validate the db has a hole here and error if not, we don't attempt to repair because something must be very wrong for this to fail
+		return si.validateIsNullRound(ctx, epoch)
+	}
+
+	// if the index is empty -> short-circuit and simply backfill if applicable
+	if isIndexEmpty {
+		if !backfill {
+			return nil, makeBackfillRequiredErr(epoch)
+		}
+		return si.backfillMissingTipset(ctx, expectedTs)
+	}
+	// see if the tipset at this epoch is already indexed or if we need to backfill
+	revertedCount, nonRevertedCount, err := si.getTipsetCountsAtHeight(ctx, epoch)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			if !backfill {
+				return nil, makeBackfillRequiredErr(epoch)
+			}
+			return si.backfillMissingTipset(ctx, expectedTs)
+		}
+		return nil, xerrors.Errorf("failed to get tipset counts at height %d: %w", epoch, err)
+	}
+
+	switch {
+	case revertedCount == 0 && nonRevertedCount == 0:
+		// no tipsets at this epoch in the index, backfill
+		if !backfill {
+			return nil, makeBackfillRequiredErr(epoch)
+		}
+		return si.backfillMissingTipset(ctx, expectedTs)
+
+	case revertedCount > 0 && nonRevertedCount == 0:
+		return nil, xerrors.Errorf("index corruption: height %d only has reverted tipsets", epoch)
+
+	case nonRevertedCount > 1:
+		return nil, xerrors.Errorf("index corruption: height %d has multiple non-reverted tipsets", epoch)
+	}
+
+	// fetch the non-reverted tipset at this epoch
+	var indexedTsKeyCidBytes []byte
+	err = si.stmts.getNonRevertedTipsetAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&indexedTsKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get non-reverted tipset at height %d: %w", epoch, err)
+	}
+
+	indexedTsKeyCid, err := cid.Cast(indexedTsKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+	expectedTsKeyCid, err := expectedTs.Key().Cid()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+	if !indexedTsKeyCid.Equals(expectedTsKeyCid) {
+		return nil, xerrors.Errorf("index corruption: indexed tipset at height %d has key %s, but canonical chain has %s", epoch, indexedTsKeyCid, expectedTsKeyCid)
```

Comment:
This, and some of the previous errors with reverted & non-reverted, suggests a very high degree of confidence that it should be nearly impossible to get into this state—are you that confident? Because at this point, similar to the `revertedCount > 0 && nonRevertedCount == 0` case, you could fairly easily fix it. Revert what's there, backfill new.

But I think by erroring you're saying that this shouldn't be possible with the new reconciliation method so the only reason you should expect this method to actually perform a backfill is if there are no tipsets already indexed at this epoch, correct?

Reply:
@rvagg Yes -> exactly. This is a very invalid state which points to something having gone drastically wrong. 

We shouldn't be repairing/backfilling this silently as it points to a deeper bug in the code.

Reply:
ok, let the come complain to us and it forces us to figure out what has gone wrong, fair